primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -652,6 +655,81 @@ where
     }
 }
 
+#[cfg(feature = "icicle")]
+impl<E: Pairing> UnivariateKzgPCS<E> {
+    /// Comupte PCS commit using GPU
+    /// Similar to [`Self::commit()`] but with ICICE's GPU-accelerated MSM
+    pub fn commit_with_gpu<C>(
+        prover_param: impl Borrow<UnivariateProverParam<E>>,
+        poly: &DensePolynomial<E::ScalarField>,
+    ) -> Result<Commitment<E>, PCSError>
+    where
+        C: IcicleCurve + MSM<C>,
+        C::ScalarField: ArkConvertible<ArkEquivalent = E::ScalarField>,
+        C::BaseField: ArkConvertible<ArkEquivalent = <C::ArkSWConfig as CurveConfig>::BaseField>,
+        E: Pairing<G1Affine = Affine<<C as IcicleCurve>::ArkSWConfig>>,
+    {
+        // TODO: use proper PCSError instead of unwrap.
+        #[cfg(feature = "kzg-print-trace")]
+        let commit_time =
+            start_timer!(|| format!("Committing to polynomial of degree {} ", poly.degree()));
+
+        let prover_param = prover_param.borrow();
+        if poly.degree() > prover_param.powers_of_g.len() {
+            return Err(PCSError::InvalidParameters(format!(
+                "poly degree {} is larger than allowed {}",
+                poly.degree(),
+                prover_param.powers_of_g.len()
+            )));
+        }
+
+        #[cfg(feature = "kzg-print-trace")]
+        let conv_time = start_timer!(|| "Type Conversion: arkworks->ICICLE");
+
+        let size = poly.degree() + 1;
+        let bases: Vec<IcicleAffine<C>> = prover_param.powers_of_g[..size]
+            .iter()
+            .map(|&p| IcicleAffine::<C>::from_ark(p))
+            .collect();
+        let scalars: Vec<C::ScalarField> = poly.coeffs()[..size]
+            .iter()
+            .map(|&s| C::ScalarField::from_ark(s))
+            .collect();
+        #[cfg(feature = "kzg-print-trace")]
+        end_timer!(conv_time);
+
+        // load them on host first
+        let bases = HostOrDeviceSlice::Host(bases);
```

Comment:
Do we need to convert and load SRS bases everytime during committing? Can we just load it once and reuse it?

Reply:
Another point: in VID, the polynomial degree is much lower. Thus we also won't need to upload too many `powers_of_g` elements to GPUs.

Reply:
I see. I feel it might be better to split it at the PCS level rather than in the VID code. Because the PCS commit function itself shouldn't upload SRS everytime. Is it possible that we add another API, `load_srs_to_gpu()`, and the `commit_in_gpu` function can take `&HostOrDevicesSlice<T>` and no longer need to upload srs, and it will return an error if SRS hasn't been uploaded?

In the VID code, it will call `load_srs_to_gpu` and `commit_in_gpu` when needed.

Reply:
your suggestion makes more sense! I'll implement that!

Reply:
a slight annoyance is lifetime due to `HostOrDeviceSlice<'a, T>` as a return parameter, I've been fighting this today. I don't want to assign `'static` for it, as we shouldn't expect the reference to live that long, we only want it to be as long as the cuda pointer being active.

I'll figure sth out, but just to share some engineering journey.

Reply:
done in [514d479](https://github.com/EspressoSystems/jellyfish/pull/498/commits/514d479ea4a17314e09e057ef6e888bb8828cf6a)